compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEFuseConsumerIntoLoop.cpp

```diff
@@ -658,80 +36,103 @@ void AMDAIEFuseConsumerIntoLoopPass::runOnOperation() {
-  auto loopBody = useSCFFor ? cast<scf::ForOp>(loops).getBody()
-                            : cast<scf::ForallOp>(loops).getBody();
+  // The depth until which we would keep fusing consumer chain.
+  // TODO(avarma): This should also be part of KernelDispatch logic.
+  unsigned fuseDepth = 1;
+  // Check if there is matmul-elementwise fusion opportunity. If so, overwrite
+  // the `fuseDepth` to be 2.
+  funcOp->walk<WalkOrder::PostOrder, ReverseIterator>([&](linalg::LinalgOp op) {
+    if (isMatmulProducerOfElementwise(op)) {
+      fuseDepth = 2;
+      return WalkResult::interrupt();
+    }
+    return WalkResult::advance();
+  });
 
-  // Search the compute op and its consumer slices.
-  linalg::LinalgOp linalgOp;
-  loopBody->walk<WalkOrder::PostOrder, ReverseIterator>(
-      [&](linalg::LinalgOp op) {
-        if (linalg::isaContractionOpInterface(op)) {
-          linalgOp = op;
-          return WalkResult::interrupt();
-        }
-        return WalkResult::advance();
-      });
+  // Based on the `fusePackDepth`, we would greedily fuse the producer
+  // tensor.pack ops.
+  for (unsigned depth = 1; depth <= fuseDepth; depth++) {
+    // Walk through the graph in post order and find the loop.
+    Operation *scfLoopOp = nullptr;
+    funcOp->walk<WalkOrder::PostOrder, ReverseIterator>(
+        [&](LoopLikeOpInterface op) {
+          if (isa<scf::ForOp>(op) && useSCFFor) {
+            scfLoopOp = op;
+            return WalkResult::interrupt();
+          } else if (isa<scf::ForallOp>(op) && !useSCFFor) {
+            scfLoopOp = op;
+            return WalkResult::interrupt();
+          }
+          return WalkResult::advance();
+        });
+
+    if (!scfLoopOp) {
+      LLVM_DEBUG(llvm::dbgs()
+                 << "There is no scf.for/forall loop to fuse with.\n");
+      return;
+    }
 
-  if (!linalgOp) {
-    LLVM_DEBUG(llvm::dbgs()
-               << "----- Could not find any linalg op implementing "
-                  "ContractionOpInterface \n");
-    return;
-  }
+    LoopLikeOpInterface loops = cast<LoopLikeOpInterface>(scfLoopOp);
+    auto loopBody = useSCFFor ? cast<scf::ForOp>(loops).getBody()
+                              : cast<scf::ForallOp>(loops).getBody();
+    // Search the compute op and its consumer slices.
+    linalg::LinalgOp linalgOp;
+    // TODO(avarma): All the following combination seems to be giving the same
+    //               results for an IR Walk :-
+    //               1. PostOrder + ReverseIterator.
+    //               2. PreOrder.
+    //               3. PreOrder + ReverseIterator.
+    //               Something while IR formation seems to be performing an ill
+    //               operation maybe?
+    loopBody->walk<WalkOrder::PostOrder, ReverseIterator>(
+        [&](linalg::LinalgOp op) {
+          if ((depth == 1 && linalg::isaContractionOpInterface(op)) ||
+              (depth == 2 && !linalg::isaContractionOpInterface(op))) {
```

Comment:
Here I don't think you need to check depth == 1 or 2 after you assign the constant value to fuseDepth.

Reply:
Actually this would indeed be required had the `WalkOrder::PostOrder, ReverseIterator` walk didn't work (which it now does after the previous fix).